### PR TITLE
Timer to post topics

### PR DESCRIPTION
**Pull request by mcass19**
When I started doing this feature, I tried to do something like [this](https://github.com/mhuggins/jquery-countTo) or [this](https://github.com/bfintal/Counter-Up), but I could not make it work by using that. Therefore, I decided to take a more creative approach and write my own timer arriving at the solution presented in _timer.html.erb.

I'm going to fix all the styles of the app in another PR so don't pay much attention on that !

### CODE REVIEW

**Comment by mcass19** [comment]
Not updated yet !!

File: app/views/retrospectives/_timer.html.erb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  const count = setInterval(function() {
+    seconds++;
+    if (seconds == 60) {
+      minutes++;
+      seconds = 0;
+    }
+    if (minutes < 10) {
+      minutesToPrint = '0' + minutes;
+    }
+    else {
```
Comment (FLarra) [inline app/views/retrospectives/_timer.html.erb]: I personally prefer writting `if else` conditionals in the way https://google.github.io/styleguide/jsguide.html does, which is 

```js
if (...) {
} else {
}
```

Don't you love it? 😍

File: app/views/retrospectives/_timer.html.erb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<script type="text/javascript">
+  const elem = $('#counter');
```
Comment (FLarra) [inline app/views/retrospectives/_timer.html.erb]: I know this case is a bit extrange. First let me share the `const` documentation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const

Second, It's weird that some lines below you are changing the html content for it. I'm not sure what is the best way to handle it in this case but I'd have used a `let` instead of a `const`. The element and properties are the same that's why it works anyways but my mind tends to think that since it is declared as a constant, you won't change it.

JS is different to any other languages in the handling of constants, so if you can investigate and convince me that this is the way to do it or that it should be a let I'd be really thankful.

File: app/views/retrospectives/_timer.html.erb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  let seconds = 0;
+  let minutes = 0;
+  let minutesToPrint;
+  let secondsToPrint;
+  const count = setInterval(function() {
```
Comment (FLarra) [inline app/views/retrospectives/_timer.html.erb]: If we needed to be completely precise in the timer, using `setInterval` would be a problem. It's not the case so we are good here, but just have that in mind. 

For further information:
https://www.quora.com/How-timers-and-intervals-work-in-the-single-threaded-nature-of-JavaScript

File: app/views/retrospectives/index.html.erb [hunk 4]
```diff
@@ -1,3 +1,3 @@
 <div class='retrospectives_index'>
-  <%= button_to 'Start new Retro!', retrospectives_path, method: 'get', class: 'button' %>
+  <%= button_to 'Start new Retro!', new_retrospective_path, method: 'get', class: 'button' %>
```
Comment (FLarra) [inline app/views/retrospectives/index.html.erb]: I think this is not related to what the commit message suggest. Keep in mind to try making commit messages the most accurate to what code changes do as possible in order to help code reviewer get context quicker

File: app/views/retrospectives/_timer.html.erb (hunk 2 above)
Comment (fchavat) [inline app/views/retrospectives/_timer.html.erb]: Maybe you can use a more descriptive name for this variable

File: app/assets/stylesheets/retrospective.scss [hunk 5]
```diff
@@ -19,3 +19,39 @@
     }
   }
 }
+
+.retrospectives_new {
+  .button {
```
[the remark this answers is not in the document]
Comment (mcass19) [reply app/assets/stylesheets/retrospective.scss]: I agree, so I've change that in https://github.com/mcass19/retro_app_mauri/pull/11.

File: app/assets/stylesheets/retrospective.scss [hunk 6]
```diff
@@ -19,3 +19,39 @@
...
+  }
+
+  .timer {
+    color: #2AB667;
+    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
+    font-size: 20px;
+    font-weight: bold; //acceptance criteria
+  }
+
+  .timer-last-minutes {
```
[the remark this answers is not in the document]
Comment (mcass19) [reply app/assets/stylesheets/retrospective.scss]: I agree, so I've change that in https://github.com/mcass19/retro_app_mauri/pull/11.

File: app/views/retrospectives/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<% content_for :add_to_header do %>
```
Comment (fchavat) [inline app/views/retrospectives/new.html.erb]: The name `add_to_headers` sounds like an event more than a place where some code will be stored, what do you think?.
Maybe a name like just `header` is more suitable?

File: Gemfile [hunk 8]
```diff
@@ -24,6 +24,8 @@ gem 'coffee-rails', '~> 4.2'
 gem 'turbolinks', '~> 5'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
 gem 'jbuilder', '~> 2.5'
+# jquery
+gem 'jquery-rails', '~> 4.3', '>= 4.3.1'
```
[the remark this answers is not in the document]
Comment (mcass19) [reply Gemfile]: There isn't 😁! I've removed the version in 723d2b8.

File: app/assets/stylesheets/retrospective.scss [hunk 9]
```diff
@@ -19,3 +19,39 @@
...
+.retrospectives_new {
+  .button {
+    color: #6E6E6E;
+    background-color: #D2D2D2;
+    padding: 5px 20px;
+    border-radius: 12px;
+    box-shadow: 0 8px 16px 0 rgba(0 , 0, 0, .2), 0 6px 20px 0 rgba(0, 0, 0, .19);
+    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
+    font-size: 15px;
+    &:hover {
```
[the remark this answers is not in the document]
Comment (mcass19) [reply app/assets/stylesheets/retrospective.scss]: I agree, so I've change that in #11.

File: app/views/retrospectives/_timer.html.erb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    seconds++;
+    if (seconds == 60) {
```
[the remark this answers is not in the document]
Comment (mcass19) [reply app/views/retrospectives/_timer.html.erb]: Great ! I forgot that javascript have `===` comparison. Thanks!

File: app/views/retrospectives/_timer.html.erb (hunk 1 above)
Comment (mcass19) [reply app/views/retrospectives/_timer.html.erb]: Mmm we can discuss about that if you want 😂. I'll take that into account !

File: app/views/retrospectives/_timer.html.erb (hunk 2 above)
Comment (mcass19) [reply app/views/retrospectives/_timer.html.erb]: I agree that it is clearer to define it with a `let` because it will change its value in the execution, going against what the definition of const specifies. I did not define it with the semantic that it was a constant, so I'm going to change it. Thank you !

File: app/views/retrospectives/index.html.erb (hunk 4 above)
Comment (mcass19) [reply app/views/retrospectives/index.html.erb]: I'm sorry for this, I think I lost this change and then forgot to put into another commit. Thanks for the comment !

File: app/views/retrospectives/new.html.erb (hunk 7 above)
Comment (mcass19) [reply app/views/retrospectives/new.html.erb]: This name is because I define a `yield :add_to_header` in the header itself, and the views that gonna change the header, add content to this. If you think it's wrong, let me know and I'll change it.

File: app/views/retrospectives/_timer.html.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (mcass19) [reply app/views/retrospectives/_timer.html.erb]: Great ! I'm on it.

**Comment by mcass19** [comment]
@FLarra @fchavat I have updated the timer code taking into account all your comments in ba19b21.

I have this warning in the browser even though I set `async: true` in the` javascript_include_tag`: 

image

I looked for how to solve it but I still have not come up with a solution. However, I prefer to continue working and in the future, if a solution is found, make the corresponding correction.

**Comment by mcass19** [comment]
@FLarra @fchavat I found a solution to that warning by making a better and more orderly solution in 71e9109:

- Add `yield` in the `<head>` of the application layout to add scripts to the views with `content_for`
- I included the javascript in the `new` view
- I put the timer code inside a function
- I call that function when the `_timer` partial is rendered

**Comment by fchavat** [comment]
Looks good!!
I have some small concerns
- Did you removed the `require_tree .` from `application.js` manifest for a reason? 
This will make the asset pipeline not include your .js files in the unified file that it generates, so, you will have to include files manually each time you need them (and loose the performance advantage of asset pipeline). I think that if you considered that and you want to keep doing this way it's okay for me 😄 
- Are you applying the fixes recognized in this pull request in a new pull request for a reason? I think that it is easier to keep track of the fixes if you commit them in the same pull request in which they were founded (for you and for the reviewers 😀 ), but maybe there is a good reason to do it this way!!

Apart of this, I think that it's all great 🎉!

**Comment by mcass19** [comment]
@fchavat thanks for the comments !

- I have removed the `require_tree .` from the` application.js` to not have the scripts in all the views. If I add this again, `timer.js` will be everywhere and will only run in one view. I do not know if it is clear and it is correct, if you do not think so and there is a better way I would be very grateful if you share it with me.

- I am not applying the corrections recognized in another PR, those changes are already made a few days ago :sweat_smile:
